### PR TITLE
makes action buttons bindable to keybinds

### PR DESCRIPTION
**Pull request by GDNgit**
## What Does This PR Do
Makes action buttons able to be binded to keybinds
Pressing CTRL-SHIFT on a action button will open a menu up that allows for a keybind to any key you input. You cannot bind an action button to multiple keys (mostly because this would just cause confusion with binding a key to the same button twice). 

## Why It's Good For The Game
Not everyone has a great mouse, you should have the option to just hit a button if you want to use an ability. Accessibility is important and we've already pretty much moved away from focusing on legacy mode. 

## Testing
I clicked quite a few buttons as a ling
## Changelog
:cl:
tweak: Action buttons can now be binded to keybinds, CTRL+SHIFT click an action button to bind a key to it
/:cl:

### CODE REVIEW

**Review by S34NW — CHANGES_REQUESTED** [review]
Please fix merge conflicts for TM viability. This PR is very stale.

**Comment by GDNgit** [comment]
> Please fix merge conflicts for TM viability. This PR is very stale.

resolved

**Comment by matttheficus** [comment]
- Multi-binding leads to some odd behavior - for example, you can pop cling armor + armblade in one keypress instantly activating both abilities.
- There's no way to unbind the keys.
- There's no way to specify numpad vs number keys.
- If possible, it would be neat if you could see a little icon/character/whatever on the action button for what the keybind is set to.
- There should be a way to blacklist certain abilities from being keybound. Not sure which yet, if any, but certain admin-only items, dust implants, Evolution/Vamp Subtype Menu, etc probably should be able to be keybound.

**Comment by GDNgit** [comment]
> * Multi-binding leads to some odd behavior - for example, you can pop cling armor + armblade in one keypress instantly activating both abilities.

You cannot, click delays are respected. It does let you retract it very quickly but that's all. 

> * There's no way to unbind the keys.

Added in latest commit, just don't enter anything and it'll clear your active bind

> * There's no way to specify numpad vs number keys.

Code limitation as of now, I'll see if I can figure a way around it
> * If possible, it would be neat if you could see a little icon/character/whatever on the action button for what the keybind is set to.

Binded key added to the mouseover of action buttons

> * There should be a way to blacklist certain abilities from being keybound. Not sure which yet, if any, but certain admin-only items, dust implants, Evolution/Vamp Subtype Menu, etc probably should be able to be keybound.

I think it's very *stupid* to bind these but I'm not sure if this is an actual issue. Anyways I guess this means suicide macros are real again.

**Comment by Lazyman310** [comment]
Did this PR also _remove_ the ability to right-click things to do things? I would be okay with both just to have to think about hotkeys a little less, but would rather right-clicking to do things stay

**Comment by GDNgit** [comment]
> Did this PR also _remove_ the ability to right-click things to do things? I would be okay with both just to have to think about hotkeys a little less, but would rather right-clicking to do things stay

No, that was a different pr

File: code/_onclick/hud/action_button.dm [hunk 1]
```diff
@@ -58,9 +63,33 @@
...
+		var/datum/keybinding/mob/trigger_action_button/triggerer = new
+		triggerer.linked_action = linked_action
+		user.client.active_keybindings[keybind_to_set_to] += list(triggerer)
+		linked_keybind = triggerer
+		triggerer.binded_to = keybind_to_set_to
+		to_chat(user, "<span class='info'>[src] has been binded to [keybind_to_set_to]!</span>")
+	else
+		if(linked_keybind)
+			clean_up_keybinds(user)
+			to_chat(user, "<span class='info'>Your active keybinding on [src] has been cleared.</span>")
```
Comment (Gaxeer) [inline code/_onclick/hud/action_button.dm:80]:
```suggestion
	else if(linked_keybind)
		clean_up_keybinds(user)
		to_chat(user, "<span class='info'>Your active keybinding on [src] has been cleared.</span>")
```

File: code/modules/input/keybindings_procs.dm [hunk 2]
```diff
@@ -12,5 +12,10 @@
 				movement_kb_dirs[key] = KBM.move_dir
 			else
 				active_keybindings[key] += list(KB)
+	if(mob) // Clients can join before world/new is setup, so we gotta mob check em
+		for(var/datum/action/action as anything in mob.actions)
+			if(action.button?.linked_keybind?.binded_to)
+				var/datum/keybinding/mob/trigger_action_button/linked_bind = action.button.linked_keybind
+				active_keybindings[linked_bind.binded_to] += list(linked_bind)
```
Comment (Gaxeer) [inline code/modules/input/keybindings_procs.dm:19]: Let's rewrite using guard clause, what do you think? 
```suggestion
	if(!mob) // Clients can join before world/new is setup, so we gotta mob check em
		return
		
	for(var/datum/action/action as anything in mob.actions)
		if(action.button?.linked_keybind?.binded_to)
			var/datum/keybinding/mob/trigger_action_button/linked_bind = action.button.linked_keybind
			active_keybindings[linked_bind.binded_to] += list(linked_bind)
```

Comment (GDNgit) [inline code/modules/input/keybindings_procs.dm:19]:
```suggestion
	if(!mob) // Clients can join before world/new is setup, so we gotta mob check em
		return active_keybindings
	for(var/datum/action/action as anything in mob.actions)
		if(action.button?.linked_keybind?.binded_to)
			var/datum/keybinding/mob/trigger_action_button/linked_bind = action.button.linked_keybind
			active_keybindings[linked_bind.binded_to] += list(linked_bind)
```

**Review by S34NW — APPROVED** [review]
No reported issues in over a week on TM.

**Pull request merged**